ffbidx package for the fast feedback indexer library

hcstadler opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

ndevenish commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,57 @@
+        - {{ compiler('cxx') }}
+        - {{ stdlib('c') }}
+        - cuda-cudart-dev {{ cuda_compiler_version }}
+        - conda-forge::pkg-config
+        - conda-forge::meson
+        - conda-forge::eigen  # header only library
```
You don't need `conda-forge::` here

hcstadler replied:
Done

hcstadler commented:
@conda-forge/help-c-cpp, ready for review

xhochy commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,57 @@
+
+{% set cuda_major = (cuda_compiler_version|default("12.0")).rpartition('.')[0] %}
+
+package:
```
Please indent with two spaces like we do in all other recipes.

hcstadler replied:
Done

xhochy commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,57 @@
+        - {{ compiler('cuda') }}
+        - {{ compiler('cxx') }}
+        - {{ stdlib('c') }}
+        - cuda-cudart-dev {{ cuda_compiler_version }}
```
This should be in host

hcstadler replied:
Done

ndevenish replied:
Are you sure about this? Don't compilers normally go in build? (because they don't need to be tied to the target architecture, and otherwise would prevent e.g. cross compilation).

Or did you just mean cuda-cudart-dev? (which I agree looks like it should be in host)

ndevenish replied:
I am pretty sure that this was a mobile-editing error, and that it was only intended to say that cuda-cudart-dev should be in host. So the compilers and `stdlib("c")` should be in build, not host.

xhochy commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,57 @@
+        - {{ stdlib('c') }}
+        - cuda-cudart-dev {{ cuda_compiler_version }}
+        - pkg-config
+        - meson
+        - eigen  # header only library
+        - python
+        - numpy
```
These four should also be in host

hcstadler replied:
Done.

xhochy commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,57 @@
+        - numpy
+    run:
+        - python
+        - {{ pin_compatible('numpy') }}
```
This should no longer be used 
```suggestion
```

hcstadler replied:
Done

xhochy commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,57 @@
+        - python
+        - {{ pin_compatible('numpy') }}
+
+test:
```
```suggestion
test:
  requires:
    - pip
  commands:
    - pip check
```

hcstadler replied:
Done

xhochy commented on recipes/ffbidx/build.sh:
```diff
@@ -0,0 +1,9 @@
+PKG_CONFIG_PATH=${BUILD_PREFIX}/share/pkgconfig:${PKG_CONFIG_PATH}
```
If you move the dependencies into host, you should no longer need this
```suggestion
```

hcstadler replied:
Done

xhochy commented on recipes/ffbidx/build.sh:
```diff
@@ -0,0 +1,9 @@
+PKG_CONFIG_PATH=${BUILD_PREFIX}/share/pkgconfig:${PKG_CONFIG_PATH}
+CUDA_HOME=${BUILD_PREFIX} meson setup --reconfigure --buildtype=release --prefix=${PREFIX} --libdir=lib -Ddefault_library=shared -Dinclude-python-api=enabled meson
```
```suggestion
CUDA_HOME=${BUILD_PREFIX} meson setup --reconfigure --buildtype=release --prefix=${PREFIX} --libdir=lib -Ddefault_library=shared -Dinclude-python-api=enabled meson $MESON_ARGS
```

hcstadler replied:
Done

hcstadler commented:
This fails on cuda_118 (which I thought I had skipped) in "Packaging cbfmt" - something I do not use for my code.

ndevenish commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,56 @@
+
+requirements:
+  host:
+    - {{ compiler('cuda') }}
+    - {{ compiler('cxx') }}
+    - {{ stdlib('c') }}
+    - meson
+    - pkg-config
```
Usually I would expect these four should be in `build:`. But I don't know if meson needs that to workaround the environment.

ndevenish replied:
Ah, meson was reviewed as needing to be in host, above, so maybe it does belong. It probably doesn't matter that much, if it doesn't export a run requirement.

xhochy replied:
It will matter for cross-compilation probably and with having meson here, you avoid having a python installation in build.

xhochy replied:
I think this got mixed up by the GitHub app. Compilers should only be in build.

hcstadler commented:
The linter fails if compiler('cxx') and stdlib('c') are not present in build.
cuda_120 fails if compiler('cuda') is not present in build.

hcstadler commented:
@conda-forge/help-c-cpp, ready for review

ndevenish commented:
- I'm pretty sure compilers should definitely not be in `host:`, otherwise I don't think cross-compilation would work. I think above request to move them to host was using an incorrect line selection.
- pkg-config shouldn't _need_ to be in host, because that's just the CLI tool, and meson should still be able to pick up the other dependencies in host when the actual `pkg-config` executable is on the PATH outside the host path. I am not excluding the possibility that meson does something weird requiring it though, so (unless review disagrees) it's probably fine to have in host, along with meson (I don't think either declare an associated run requirement).
- The top two `cuda_home` lines in build.sh look as though they aren't doing anything?

hcstadler commented:
@ndevenish: I agree with you on that.
The cuda_home lines in build.sh compute the cuda home path (for a typical installation) from the location of the nvcc binary, which in a typical installation is under bin/nvcc of the cuda home path. If I don't set CUDA_HOME, the meson build fails as far as I remember.
Much of my and mesons confusion stems from the fact that compiler('cuda') results in fundamentally different cuda installations for cuda_118 and cuda_120. I decided to only support cuda_120.

dagewa commented:
This came up in our DIALS core meeting today. I'm a bit confused about the status, are we waiting for re-review by @xhochy?

xhochy commented:
This us waiting for the removal of the compilers from host

carterbox commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+  skip: True  # [(cuda_compiler_version or "None") == "None"]
+  skip: True  # [cuda_compiler_version == "11.8"]
+
+requirements:
+  build:
+    - {{ compiler('cuda') }}
+    - {{ compiler('cxx') }}
+    - {{ stdlib('c') }}
+  host:
+    - {{ compiler('cuda') }}
+    - {{ compiler('cxx') }}
+    - {{ stdlib('c') }}
+    - meson
+    - pkg-config
+    - eigen  # header only library
+    - python
+    - numpy
```
```suggestion
requirements:
  build:
    - {{ compiler('cuda') }}
    - {{ compiler('cxx') }}
    - {{ stdlib('c') }}
    - meson
    - pkg-config
  host:
    - eigen  # header only library
    - python
    - numpy
```

carterbox commented on recipes/ffbidx/meta.yaml:
```diff
@@ -0,0 +1,55 @@
+build:
+  number: 0
+  skip: True  # [not linux64]
+  skip: True  # [(cuda_compiler_version or "None") == "None"]
+  skip: True  # [cuda_compiler_version == "11.8"]
```
```suggestion
  skip: True  # [cuda_compiler_version in (None, "None", "11.8")]
```

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/ffbidx/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipes/ffbidx/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. Your recipe may not receive automatic updates and/or may not be compatible with conda-forge's infrastructure. Please check the logs for more information and ensure your recipe can be parsed.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/11999244239](https://github.com/conda-forge/conda-forge-webservices/actions/runs/11999244239). Examine the logs at this URL for more detail.</sub>

carterbox commented on recipes/ffbidx/build.sh:
```diff
@@ -0,0 +1,10 @@
+cuda_home=$(which nvcc)
+cuda_home=${cuda_home%/bin/nvcc}
+CUDA_HOME=${cuda_home} meson setup $MESON_ARGS --reconfigure -Ddefault_library=shared -Dinclude-python-api=enabled meson
```
Looks like you neeed to add some args to help meson find NumPy in the host prefix.

carterbox reviewed (changes requested):
If you look at the logs, you will see that the python versions in the build ($BUILD_PREFIX) and host ($PREFIX) environment are different and that meson detects the wrong one (the one in build instead of host). This is why I asked you to add some extra args to your build script to help meson find the python/numpy from the host prefix ($PREFIX).

hcstadler commented:
@carterbox: Thanks for the great review and suggestions. I added an option to select the python installation in my meson build and updated the recipe, so the python installation to build against is now set to the one installed under $PREFIX.

hcstadler commented:
@conda-forge/help-c-cpp, ready for review

xhochy commented:
Had to merge main here as the build was building a different recipe than the one in the PR.
